src/services/services.ts

```diff
@@ -6459,6 +6634,268 @@ namespace ts {
+                        return getReferenceEntryFromNode(refNode.parent);
+                    }
+                    else if (refNode.parent.kind === SyntaxKind.ShorthandPropertyAssignment) {
+                        // Go ahead and dereference the shorthand assignment by going to its definition
+                        return getReferenceEntryForShorthandPropertyAssignment(refNode, typeChecker);
+                    }
+
+                    // Check if the node is within an extends or implements clause
+                    const containingHeritageClause = getContainingClassHeritageClause(refNode);
+                    if (containingHeritageClause) {
+                        return getReferenceEntryFromNode(containingHeritageClause.parent);
+                    }
+
+                    // If we got a type reference, try and see if the reference applies to any expressions that can implement an interface
+                    const containingTypeReference = getContainingTypeReference(refNode);
+                    if (containingTypeReference) {
+                        const parent = containingTypeReference.parent;
+                        if (isVariableLike(parent) && parent.type === containingTypeReference && parent.initializer && isImplementationExpression(parent.initializer)) {
+                            return getReferenceEntryFromNode(parent.initializer);
+                        }
+                        else if (isFunctionLike(parent) && parent.type === containingTypeReference && parent.body && parent.body.kind === SyntaxKind.Block) {
+                            return forEachReturnStatement(<Block>parent.body, (returnStatement) => {
+                                if (returnStatement.expression && isImplementationExpression(returnStatement.expression)) {
+                                    return getReferenceEntryFromNode(returnStatement.expression);
+                                }
+                            });
+                        }
+                        else if (isTypeAssertionExpression(parent) && isImplementationExpression(parent.expression)) {
+                            return getReferenceEntryFromNode(parent.expression);
+                        }
+                    }
+                }
+            }
+
+            function isMemberOfSubOrParentClass(reference: Node, base: Type, hierarchy: Symbol[]) {
+                if (!base || !hierarchy) {
+                    return true;
+                }
+                const referenceSymbol = typeChecker.getSymbolAtLocation(reference);
+                if (referenceSymbol && referenceSymbol.parent) {
+                    if (hierarchy.indexOf(referenceSymbol.parent) !== -1) {
+                        return true;
+                    }
+
+                    const referenceParentDeclarations = referenceSymbol.parent.getDeclarations();
+                    if (referenceParentDeclarations.length) {

```

Comment:
you should be able to check `referenceSymbol.parent.valueDeclaration` here instead of length and first declaration